src/job_token.rs

```diff
@@ -1,139 +1,160 @@
-    // but it's an extra dependency that will almost never be used, so
-    // it's generally not too worth it.
-    let mut parallelism = 4;
-    if let Ok(amt) = env::var("NUM_JOBS") {
-        if let Ok(amt) = amt.parse() {
-            parallelism = amt;
+    impl JobServer {
+        pub(super) unsafe fn from_env() -> Option<Self> {
+            let var = var_os("CARGO_MAKEFLAGS")
+                .or_else(|| var_os("MAKEFLAGS"))
+                .or_else(|| var_os("MFLAGS"))?;
+
+            let inner = sys::JobServerClient::open(var)?;
+
+            Some(Self {
+                inner,
+                global_implicit_token: AtomicBool::new(true),
+            })
+        }
+
+        pub(super) fn try_acquire(&self) -> Result<Option<JobToken>, Error> {
+            if !self.global_implicit_token.swap(false, Relaxed) {
+                // Cold path, no global implicit token, obtain one
+                if self.inner.try_acquire()?.is_none() {
+                    return Ok(None);
+                }
+            }
+            Ok(Some(JobToken()))
+        }
+
+        pub(super) fn release_token_raw(&self) {
+            if self
+                .global_implicit_token
+                .compare_exchange(false, true, Relaxed, Relaxed)
+                .is_err()
+            {
+                // There's already a global implicit token, so this token must
+                // be released back into jobserver
+                let _ = self.inner.release();
+            }
         }
     }
+}
+
+mod inprocess_jobserver {
+    use super::JobToken;
+
+    use std::{
+        env::var,
+        sync::atomic::{AtomicU32, Ordering::Relaxed},
+    };
+
+    pub(crate) struct JobServer(AtomicU32);
 
-    // If we create our own jobserver then be sure to reserve one token
-    // for ourselves.
-    let client = jobserver::Client::new(parallelism).expect("failed to create jobserver");
-    client.acquire_raw().expect("failed to acquire initial");
-    return client;
+    impl JobServer {
+        pub(super) fn new() -> Self {
+            // Use `NUM_JOBS` if set (it's configured by Cargo) and otherwise
+            // just fall back to a semi-reasonable number.
+            //
+            // Note that we could use `num_cpus` here but it's an extra
+            // dependency that will almost never be used, so
+            // it's generally not too worth it.
+            let mut parallelism = 4;
+            if let Ok(amt) = var("NUM_JOBS") {
+                if let Ok(amt) = amt.parse() {
+                    parallelism = amt;
+                }
+            }
+
+            Self(AtomicU32::new(parallelism))
+        }
+
+        pub(super) fn try_acquire(&self) -> Option<JobToken> {
+            let res = self.0.fetch_update(Relaxed, Relaxed, |tokens| {
+                if tokens > 0 {
+                    Some(tokens - 1)
+                } else {
+                    None
+                }
```

Comment:
```suggestion
                tokens.checked_sub(1)
```

Reply:
`token` is an atomic and it doesn't have `checked_sub`, using `fetch_update` is the most simple solution I can find.

Reply:
I'm not suggesting to change the `fetch_update`; though `tokens` in this function is a `u32`, so we can use `checked_sub`.

Reply:
Thanks!
Sorry for the misunderstanding

Reply:
No worries =)